libp2p/protocols/pubsub/floodsub.nim

```diff
@@ -192,41 +193,34 @@ method init*(f: FloodSub) =
   f.handler = handler
   f.codec = FloodSubCodec
 
-method publish*(
+method doPublish*(
     f: FloodSub, topic: string, data: seq[byte]
-): Future[int] {.async: (raises: []).} =
+): Future[PublishResult] {.async: (raises: []).} =
   # base returns always 0
-  discard await procCall PubSub(f).publish(topic, data)
+  discard await procCall PubSub(f).doPublish(topic, data)
 
   trace "Publishing message on topic", data = data.shortLog, topic
 
   if topic.len <= 0: # data could be 0/empty
     debug "Empty topic, skipping publish", topic
-    return 0
+    return err(NoTopicSpecified)
 
   let peers = f.floodsub.getOrDefault(topic)
 
   if peers.len == 0:
     debug "No peers for topic, skipping publish", topic
-    return 0
-
-  let
-    msg =
-      if f.anonymize:
-        Message.init(none(PeerInfo), data, topic, none(uint64), false)
-      else:
-        inc f.msgSeqno
-        Message.init(some(f.peerInfo), data, topic, some(f.msgSeqno), f.sign)
-    msgId = f.msgIdProvider(msg).valueOr:
-      trace "Error generating message id, skipping publish", error = error
-      return 0
+    return err(NoPeersToPublish)
+
+  let (msg, msgId) = f.createMessage(topic, data).valueOr:
+    trace "Error creating message, skipping publish", error = error
+    return err(CannotGenerateMessageId)
 
   trace "Created new message", payload = shortLog(msg), peers = peers.len, topic, msgId
 
   if f.addSeen(f.salt(msgId)):
     # custom msgid providers might cause this
     trace "Dropping already-seen message", msgId, topic
-    return 0
+    return err(DuplicateMessage)
```

Comment:
as discussed previously, this is not an _error_ - it's entirely normal in gossip networks - if the protocol that relies on gossupsub makes any assumptions about this point, _that protocol_ is likely flawed, ie it should not be designed around this kind of assumptions